### ISSUE TITLE
[all tools] hang with 100% CPU w/ Open MPI 4.0.x

### ISSUE THREAD

**Issue by kcgthb** [hpc/mpifileutils#302]
Hi, 

All the tools (`dwalk`, `dcp`, `dcmp`... from release 0.9.1) hang at start with all processes using 100% of the allocated CPUs. No starting message is displayed, and all the MPI ranks infinitely loop on `poll()`.

Here's a typical command:
```
$ ompi_info --version | head -n1
Open MPI v4.0.1
$ srun -N 2 --ntasks-per-node 8 dcmp -v $SRC $DEST
srun: job 46917263 queued and waiting for resources
srun: job 46917263 has been allocated resources

```
Nothing is ever displayed, and on one of the nodes, the `dcmp` processes look like this:
```
s: 417 total,   9 running, 408 sleeping,   0 stopped,   0 zombie
%Cpu(s): 49.6 us,  0.4 sy,  0.0 ni, 50.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem : 65772240 total, 47135936 free,  2413552 used, 16222752 buff/cache
KiB Swap:  4194300 total,  4194300 free,        0 used. 60403040 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 2805 kilian    20   0  702944  78320   9340 R 100.0  0.1  11:47.96 dcmp
 2806 kilian    20   0  571860  78304   9324 R 100.0  0.1  11:44.96 dcmp
 2807 kilian    20   0  584152  88540   9316 R 100.0  0.1  11:48.18 dcmp
 2809 kilian    20   0  571860  78288   9312 R 100.0  0.1  11:46.38 dcmp
 2810 kilian    20   0  571860  78296   9316 R 100.0  0.1  11:47.30 dcmp
 2811 kilian    20   0  584152  88544   9316 R 100.0  0.1  11:48.06 dcmp
 2812 kilian    20   0  571860  78280   9308 R 100.0  0.1  11:45.82 dcmp
 2808 kilian    20   0  584152  88548   9320 R  93.8  0.1  11:44.41 dcmp
 4026 kilian    20   0  161028   2368   1516 R   6.2  0.0   0:00.01 top
 3054 kilian    20   0  162156   3044   1044 S   0.0  0.0   0:00.05 sshd
 3056 kilian    20   0  121428   9328   1708 S   0.0  0.0   0:00.12 bash
```
Stracing one of them shows this:
```
[kilian@sh-09-06 ~]$ strace -Ttt -p 2805
strace: Process 2805 attached
14:47:41.697393 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000020>
14:47:41.707480 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.717547 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.727610 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.737666 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.747721 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.757779 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.767824 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.777880 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.787942 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.798034 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.808100 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.818154 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.828220 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.838282 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.848341 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.858389 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000022>
14:47:41.868450 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
14:47:41.878514 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000016>
14:47:41.888564 poll([{fd=5, events=POLLIN}, {fd=17, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000017>
```

This only happens with Open MPI 4.0.x. The exact same binary used with Open MPI 3.x works perfectly. 
In other words, just changing the Open MPI version used at runtime without even re-compiling, running the exact same command on the same nodes, works perfectly:
```
$ ompi_info --version | head -n1
Open MPI v3.1.2
$ srun -N 2 --ntasks-per-node 8 dcmp -v $SRC $DST
srun: job 46917115 queued and waiting for resources
srun: job 46917115 has been allocated resources
[2019-07-26T14:30:04] Walking source path
[2019-07-26T14:30:04] Walking /scratch/src
[2019-07-26T14:30:05] Walked 97703 items in 1.197265 seconds (81605.180128 files/sec)
[2019-07-26T14:30:05] Walking destination path
[2019-07-26T14:30:05] Walking /scratch/dest
[2019-07-26T14:30:08] Walked 97703 items in 3.192001 seconds (30608.701139 files/sec)
[2019-07-26T14:31:52] Started   : Jul-26-2019, 14:30:08
[2019-07-26T14:31:52] Completed : Jul-26-2019, 14:31:52
[2019-07-26T14:31:52] Seconds   : 104.136
[2019-07-26T14:31:52] Items     : 97703
[2019-07-26T14:31:52] Item Rate : 97703 items in 104.135739 seconds (938.227362 items/sec)
[2019-07-26T14:31:52] Bytes read: 2.212 GB (2375423896 bytes)
[2019-07-26T14:31:52] Byte Rate : 21.754 MB/s (2375423896 bytes in 104.136 seconds)
Number of items that exist in both directories: 97703 (Src: 97703 Dest: 97703)
Number of items that exist only in one directory: 0 (Src: 0 Dest: 0)
Number of items that exist in both directories and have the same type: 97703 (Src: 97703 Dest: 97703)
Number of items that exist in both directories and have different types: 0 (Src: 0 Dest: 0)
Number of items that exist in both directories and have the same content: 97703 (Src: 97703 Dest: 97703)
Number of items that exist in both directories and have different contents: 0 (Src: 0 Dest: 0)
```

Do you have any insight on what could cause that behavior with Open MPI 4.0.x?
Thanks!

**Comment by adammoody** [comment]
Thanks for reporting this problem, @kcgthb .  I'll see if I can reproduce this.

Can you capture a stack trace of the processes at the point of the hang, with something like [STAT](https://github.com/LLNL/STAT) or gdb?  STAT would let you grab the stack trace from all procs, but getting the trace from a single process with gdb might provide a good clue.

Are you able to reproduce the hang when reducing the number of MPI processes?

**Comment by kcgthb** [comment]
Hi @adammoody, thanks for getting back to me.

It looks like the hang only happens over multiple nodes, independently of the number of ranks (ie., 16 ranks on a single node work fine, but 2 ranks across 2 hosts don't).

I'll look into STAT, but the number of dependencies may require more time for us to give it a try. 

In the meantime, I gave a shot at `gdb`, and it pinpointed to UCX functions. I've looked deeper, and upon further investigation, it looks like we're hitting https://github.com/open-mpi/ompi/issues/6546

So it doesn't look like a mpifileutils issue after all, but rather an Open MPI / UCX one. So, sorry for the noise, I guess. :)
